covid19-dashboard/src/DataTable.tsx

```diff
@@ -0,0 +1,230 @@
+
+type OrientationType = 'desc' | 'asc';
+
+type TableStateType = {
+  sortBy: string;
+  order: OrientationType;
+  chunkIndex: number;
+};
+
+type TablePropsType = {
+  data: ValueType[];
+};
+
+/**
+ * The data table with all the essential information.
+ * @var state.sortBy: the default sortBy category.
+ * @var state.order: display data ascending or descending order
+ * @var state.chunkIndex: what chunk of data to display.
+ * @param props.data: an object where each key is a geoId
+ * and each value contains different timeSeries .
+ * The table is divided into pages.
+ */
+export default class DataTable extends React.Component<
+  TablePropsType,
+  TableStateType
+  > {
+  state = {
+    sortBy: Configuration.tableDefaultSortBy,
+    order: 'desc' as OrientationType,
+    chunkIndex: 0,
+  };
+
+  chunkSize = 30;
+  chunkedData: ValueType[][] = [];
+
+  /**
+   * Determines what type of arrow display
+   * next to the table's header/category. Or whether none at all.
+   * @param sortBy: the observed header/category.
+   */
+  sortArrow(sortBy: string) {
+    // If the input sortBy is the same as the state's sortBy,
+    // it means we are actively sorting by that header.
+    const active = sortBy === this.state.sortBy;
+
+    if (!active) {
+      return '';
+    } else if (this.state.order === 'desc') {
+      return '▼ ';
+    } else if (this.state.order === 'asc') {
+      return '▲ ';
+    }
+  }
+
+  /**
+   * When triggered, the current sorting orientation will be changed
+   * to the opposite orientation.
+   */
+  changeOrientation = () => {
+    // From ascending to descending.
+    // From descending to ascending.
+    const opposites = {asc: 'desc', desc: 'asc'};
+
+    // Change the current orientation the opposite orientation.
+    const newOrientation = opposites[this.state.order] as OrientationType;
+    this.setState({
+      order: newOrientation,
+    });
+  };
+
+  /**
+   * When tapping on the header, if the sortBy is already active,
+   * then change the orientation, otherwise, change the sortBy
+   * to the newly selected header.
+   * @param sortBy: the header/category to sort the table by.
+   */
+  sortBy(sortBy: string) {
+    const active = this.state.sortBy === sortBy;
+
+    if (active) {
+      this.changeOrientation();
+    } else {
+      this.setState({sortBy: sortBy});
+    }
+
+    // Go back to the first page in the table.
+    this.setState({chunkIndex: 0});
+  }
+
+  /**
+   * Go to a specific page index in the table.
+   * @param chunkIndex: the index to navigate to.
+   * Corresponds to the table's page - 1.
+   */
+  setChunkIndex = (chunkIndex: number) => {
+    // Make sure our table doesn't go out of boundaries.
+    if (chunkIndex >= 0 && chunkIndex < this.chunkedData.length) {
+      this.setState({chunkIndex: chunkIndex});
+    }
+  };
+
+  render() {
+    // Every time the order or sortBy is changed,
+    // we have to recalculate the new value.
```

Comment:
Only the ordering should be updated, I don't understand why values need to be recalculated? Perhaps the abstraction here needs to be updated. Each row should be a component, and you should only need to reorder the rows, and update the display to show the rows on the current page.

Reply:
Fixed, now it only sorts using a function(a, b).